app/services/lesson_completion_data.rb

```diff
@@ -0,0 +1,72 @@
+class LessonCompletionData
+  def initialize(course)
+    @course = course
+    @lessons = course.lessons
+  end
+
+  def all_completion_data
+    completion_data(percentage: true, duration: true, course_duration: true, completion_count: true)
+  end
+
+  def completion_data(data_options)
+    course_data = {}
+    course_data['course_duration'] = known_lesson_times_total.inspect if data_options[:course_duration]
+    lesson_duration_pairs = known_completion_durations
+    course_lessons_data = lesson_duration_pairs.map do |lesson, duration|
+      lesson_data = {}
+      lesson_data['percentage'] = duration_percentage(duration) if data_options[:percentage]
+      lesson_data['duration'] = duration.inspect if data_options[:duration]
+      lesson_data['users who finished'] = lesson_completions_count[lesson.id] if data_options[:completion_count]
+      [lesson.title, lesson_data]
+    end.to_h
+    course_data.merge(course_lessons_data)
+  end
+
+  private
+
+  attr_reader :course
+
+  def lesson_and_avg_completion_date_pairs
+    @lesson_and_avg_completion_date_pairs ||= LessonCompletion.all.group(:lesson_id).average('extract(epoch from created_at)')
+  end
+
+  def ordered_lessons
+    @ordered_lessons ||= @course.sections.includes(:lessons).map do |section|
+      section.lessons.sort_by(&:position)
+    end.flatten
+  end
+
+  def known_completion_durations
+    lessons = ordered_lessons
+    current_lesson = lessons[0]
+    current_lesson_avg_completion_time = lesson_and_avg_completion_date_pairs[current_lesson.id]
+    completion_durations = {}
+    lessons.each do |lesson|
+      next_lesson = lesson
+      next_lesson_avg_completion_time = lesson_and_avg_completion_date_pairs[next_lesson.id]
+      if next_lesson_avg_completion_time.nil? # if noone completed the assignment return data
+        break
+      else
+        average_duration_to_finish_next_lesson = next_lesson_avg_completion_time - current_lesson_avg_completion_time
+        completion_durations[next_lesson] = ActiveSupport::Duration.build(average_duration_to_finish_next_lesson)
+        current_lesson_avg_completion_time = next_lesson_avg_completion_time
+        current_lesson = next_lesson
+      end
+    end
+    completion_durations
+  end
+
+  def lesson_completions_count
+    lesson_completions_count ||= LessonCompletion.group(:lesson_id).count
```

Comment:
What is supposed to be happening here?

Reply:
> What is supposed to be happening here?
This is a method I added in to show the number of lesson_completions for each lesson.
it will output a hash like {43 => 213}(43 being the lesson_id)(231 being the amount of people who finished the lesson)
the ||= is there so it doesn't run the query twice.

Reply:
> can you provide a little more detail on what data you're grabbing and comparing here?

Sure thing. So This Calculates the average time between a user marking one lesson as complete and a user marking the next as complete. I can find this data from the "created_at" column of the lesson_completions table.
So this became the problem (How can I find the overall average distance between two sets of data(lesson_one.lesson_completions.created_at and the next_lessons lesson_completions.created_at)). I found the most efficient way to find this data is to find the overall averages of both sets and find the distance between the two overall averages.
` LessonCompletion.all.group(:lesson_id).average('extract(epoch from created_at)')`
This code finds the average datetime every lesson was completed in epoch format.
To find the average time a lesson takes to complete
1. Need to find the average completion DateTime of the lesson before
2. You need to find the average completion DateTime of the lesson you are calculating.
3. Need to find the difference between these two numbers
the method 
`known_completion_durations` takes the average completion date and finds the duration of each lesson (untill there are no lesson_completions for the next lessons(hence the name known))
Ask more questions if you don't full understand. I will be happy to answer.

Reply:
![user_data](https://user-images.githubusercontent.com/54362821/79586002-9159f000-809e-11ea-9eb4-f06d3b5f7cf8.PNG)
Here is a screenshot showing the output with the new data

Reply:
> 
> 
> > What is supposed to be happening here?
> 
> This is a method I added in to show the number of lesson_completions for each lesson.
> it will output a hash like {43 => 213}(43 being the lesson_id)(231 being the amount of people who finished the lesson)
> the ||= is there so it doesn't run the query twice.

Why do you need the local variable?